### PR TITLE
Implement input pipeline with tf.data API

### PR DESCRIPTION
**Pull request by reuben**
Creating a PR to run tests and see how this does in its current state.

### CODE REVIEW

File: util/config.py [hunk 1]
```diff
@@ -120,14 +120,15 @@ def initialize_globals():
         with tf.device('/job:ps/task:%d' % i):
             done_queues.append(tf.FIFOQueue(1, tf.int32, shared_name=('queue%i' % i)))
 
-    # Placeholder to pass in the worker's index as token
-    c.token_placeholder = tf.placeholder(tf.int32)
+    if not tf.executing_eagerly():
```
Comment (reuben) [inline util/config.py]: Here to allow testing in eager mode more easily.

File: native_client/BUILD [hunk 2]
```diff
@@ -85,14 +77,12 @@ tf_cc_shared_object(
     linkopts = select({
         "//tensorflow:darwin": [],
         "//tensorflow:linux_x86_64": LINUX_LINKOPTS,
-        "//tensorflow:rpi3": LINUX_LINKOPTS + ["-l:libstdc++.a"],
-        "//tensorflow:rpi3-armv8": LINUX_LINKOPTS + ["-l:libstdc++.a"],
```
Comment (reuben) [inline native_client/BUILD]: Please ignore this change, it's for local testing.

File: native_client/BUILD [hunk 3]
```diff
@@ -3,7 +3,7 @@
 load("@org_tensorflow//tensorflow:tensorflow.bzl",
      "tf_cc_shared_object", "if_cuda")
 
-load("@org_tensorflow//tensorflow/contrib/lite:build_def.bzl",
+load("@org_tensorflow//tensorflow/lite:build_def.bzl",
```
Comment (reuben) [inline native_client/BUILD]: Please ignore this change, it's for local testing.

File: native_client/BUILD [hunk 4]
```diff
@@ -85,14 +77,12 @@ tf_cc_shared_object(
...
         "//tensorflow:linux_x86_64": LINUX_LINKOPTS,
-        "//tensorflow:rpi3": LINUX_LINKOPTS + ["-l:libstdc++.a"],
-        "//tensorflow:rpi3-armv8": LINUX_LINKOPTS + ["-l:libstdc++.a"],
         "//tensorflow:windows": [],
         "//conditions:default": []
     }) + tflite_linkopts(),
     deps = select({
         "//native_client:tflite": [
-            "//tensorflow/contrib/lite/kernels:builtin_ops",
+            "//tensorflow/lite/kernels:builtin_ops",
```
Comment (reuben) [inline native_client/BUILD]: Please ignore this change, it's for local testing.

File: native_client/deepspeech.cc [hunk 5]
```diff
@@ -19,12 +19,10 @@
   #include "tensorflow/core/platform/env.h"
   #include "tensorflow/core/util/memmapped_file_system.h"
 #else // USE_TFLITE
-  #include "tensorflow/contrib/lite/model.h"
-  #include "tensorflow/contrib/lite/kernels/register.h"
+  #include "tensorflow/lite/model.h"
+  #include "tensorflow/lite/kernels/register.h"
```
Comment (reuben) [inline native_client/deepspeech.cc]: Please ignore this change, it's for local testing.

File: native_client/deepspeech.cc [hunk 6]
```diff
@@ -568,8 +606,6 @@ DS_CreateModel(const char* aModelPath,
   *retval = model.release();
   return DS_ERR_OK;
 #else // USE_TFLITE
-  TfLiteStatus status;
```
Comment (reuben) [inline native_client/deepspeech.cc]: Unused variable.

File: util/feeding.py [hunk 7]
```diff
@@ -1,198 +1,94 @@
...
+    def batch_fn(features, features_len, transcripts):
+        features = tf.data.Dataset.zip((features, features_len))
+        features = features.padded_batch(batch_size,
+                                         padded_shapes=([None, Config.n_input], []))
+        transcripts = transcripts.batch(batch_size)
+        return tf.data.Dataset.zip((features, transcripts))
+
+    dataset = (tf.data.Dataset.zip((filenames, transcripts))
+                              .map(file_to_features, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+                              .cache(cache_path)
```
Comment (reuben) [inline util/feeding.py]: If `cache_path` is not provided, caches to memory.

File: util/feeding.py [hunk 8]
```diff
@@ -1,198 +1,94 @@
...
+                              .window(batch_size, drop_remainder=True).flat_map(batch_fn)
```
Comment (reuben) [inline util/feeding.py]: The `dataset.window(batch_size).flat_map(batch_fn)` is the recommended approach for applying different batching logic to different components of the dataset. In this case, the dense tensors need padded batching, while the sparse tensor needs just batching. IMO this is a deficiency in the `padded_batch` API, it should just transparently apply sparse batching to any SparseTensors in the dataset, but for now this is how to do it.

See https://github.com/tensorflow/tensorflow/issues/18302

**Comment by reuben** [comment]
This code should not change much between now and landing, so reviews can start already. This PR needs the r1.13 update to be merged, and since that's going slowly I'm asking for reviews already to speed things up. I'm not super happy with the way `evaluate.py` turned out, so that file may change a bit, but I won't force push until reviews are done.

**Comment by reuben** [comment]
(Requesting review now but I don't expect you to do it during the weekend! :P)

**Comment by kdavis-mozilla** [comment]
Seems like the training run 4290 that's testing this is having some problems. It looks like all training epochs have inf loss through the validation losses look reasonable.

**Comment by reuben** [comment]
I'm gonna check if that problem is due to the data, this tf.data code, or cudnn_rnn. I suspect it's unrelated to this PR.

File: util/feeding.py [hunk 9]
```diff
@@ -1,198 +1,97 @@
...
+
+    dataset = (tf.data.Dataset.from_generator(generate_values,
+                                              output_types=(tf.string, (tf.int64, tf.int32, tf.int64)))
+                              .map(entry_to_features, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+                              .cache(cache_path)
+                              .window(batch_size, drop_remainder=True).flat_map(batch_fn)
+                              .prefetch(num_gpus)
+                              .repeat())
 
+    return dataset, num_batches
```
Comment (tilmankamp) [inline util/feeding.py]: I think it's time to rename this file, as it is not "feeding" anymore.

File: util/config.py (hunk 1 above)
Comment (tilmankamp) [reply util/config.py]: What are the reasons for not moving to eager execution entirely?
I'd even vote for ditching MonitoreTrainingSession centric code and the coordinator.

Comment (reuben) [reply util/config.py]: Mostly to keep changes smaller so that they can be merged more easily, but also because moving to Eager is a much more involved task, we can't just flip the switch right now. The training loop has to be restructured significantly. I plan to work on that in the near/medium term.

File: util/feeding.py (hunk 9 above)
Comment (reuben) [reply util/feeding.py]: @tilmankamp @kdavis-mozilla any suggestions? It's still handling feeding data into the network in my mind, I don't know what else to call it or in what direction it needs clarification.

**Comment by reuben** [comment]
> I'm gonna check if that problem is due to the data, this tf.data code, or cudnn_rnn. I suspect it's unrelated to this PR.

Job 4298 running into the same problem, without the cudnn_rnn changes. So it's either this code or the data...

File: DeepSpeech.py [hunk 10]
```diff
@@ -18,13 +18,11 @@
...
 from tensorflow.python.tools import freeze_graph
-from util.audio import audiofile_to_input_vector
 from util.config import Config, initialize_globals
 from util.coordinator import TrainingCoordinator
-from util.feeding import DataSet, ModelFeeder
+from util.feeding import create_dataset, samples_to_mfccs, audiofile_to_features
 from util.flags import create_flags, FLAGS
 from util.logging import log_info, log_error, log_debug, log_warn
-from util.preprocess import preprocess
 from util.text import Alphabet
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Is `Alphabet` used in this file?

File: DeepSpeech.py [hunk 11]
```diff
@@ -55,25 +53,82 @@ def variable_on_worker_level(name, shape, initializer):
...
-    The variables named ``bn``, where ``n`` is an integer, hold the learned bias variables.
-    In particular, the first variable ``h1`` holds the learned weight matrix that
-    converts an input vector of dimension ``n_input + 2*n_input*n_context``
-    to a vector of dimension ``n_hidden_1``.
-    Similarly, the second variable ``h2`` holds the weight matrix converting
-    an input vector of dimension ``n_hidden_1`` to one of dimension ``n_hidden_2``.
-    The variables ``h3``, ``h5``, and ``h6`` are similar.
-    Likewise, the biases, ``b1``, ``b2``..., hold the biases for the various layers.
-    '''
+def create_overlapping_windows(batch_x):
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Nice!

File: DeepSpeech.py [hunk 12]
```diff
@@ -329,7 +336,7 @@ def log_variable(variable, gradient=None):
     It logs scalar values for the mean, standard deviation, minimum and maximum.
     Furthermore it logs a histogram of its state and (if given) of an optimization gradient.
     '''
-    name = variable.name
+    name = variable.name.replace(':', '_')
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: Why this change?

File: DeepSpeech.py [hunk 13]
```diff
@@ -834,6 +843,7 @@ def do_graph_freeze(output_file=None, output_node_names=None, variables_blacklis
         except RuntimeError as e:
             log_error(str(e))
 
+
```
Comment (kdavis-mozilla) [inline DeepSpeech.py]: nit extra line

File: requirements.txt [hunk 14]
```diff
@@ -1,19 +1,23 @@
-pandas
-progressbar2
-python-utils
+# Main training requirements
```
[the remark this answers is not in the document]
Comment (reuben) [reply requirements.txt]: Yes. `tables` was only used by the feature caching code and for `python-utils` I couldn't find any uses.

File: util/feeding.py [hunk 15]
```diff
@@ -1,198 +1,97 @@
...
+from tensorflow.contrib.framework.python.ops import audio_ops as contrib_audio
+from util.config import Config
+from util.text import text_to_char_array
+
+
+def read_csvs(csv_files):
+    source_data = None
+    for csv in csv_files:
+        file = pandas.read_csv(csv, encoding='utf-8', na_filter=False)
+        #FIXME: not cross-platform
```
[the remark this answers is not in the document]
Comment (reuben) [reply util/feeding.py]: I don't think we support *training* on Windows, which is where this code/comment has any effect.

File: DeepSpeech.py (hunk 12 above)
Comment (reuben) [reply DeepSpeech.py]: ':' is an illegal character in TensorBoard summary names so it produces a ton of warnings.

Comment (reuben) [reply DeepSpeech.py]: (I described the reasoning in the commit message)

File: native_client/BUILD (hunk 4 above)
Comment (reuben) [reply native_client/BUILD]: Yes, when I rebase this PR on top of the r1.13 changes all of this will be gone.

File: DeepSpeech.py (hunk 13 above)
Comment (reuben) [inline DeepSpeech.py]: This was on purpose to match PEP 8 (two empty lines between top-level functions)

File: util/feeding.py (hunk 7 above)
Comment (reuben) [reply util/feeding.py]: Yep.

**Comment by reuben** [comment]
Rebased on top of r1.13 changes in master and addressed review comments in 767e0b1.

**Comment by reuben** [comment]
Two samples are causing the `inf` training loss problem:

|wav_filename|transcript|
|-|-|
|fisher-2004-split-wav/fe_03_00250-585.19-585.91.wav|yeah i didn't like it much actually|
|fisher-2005-split-wav/fe_03_05882-158.6-159.56.wav|you know what would you do if you had a million|

Before this PR, the corresponding audio files would create 36 and 48 time steps, respectively. Because of different rounding, the new feature computation generates 35 and 47 time steps instead, which is also exactly the length of their transcripts. The TF AudioSpectrogram op only creates full windows, dropping samples if needed, whereas `python_speech_features` used the entire audio by padding the last window if needed.

You can reproduce this problem with the code on master by downloading the audio files and training with the following train CSV and `--n_hidden 2048`:

```
wav_filename,wav_filesize,transcript
fe_03_00250-585.19-585.91.wav,0,yeah i didn't like it much actuallyy
fe_03_05882-158.6-159.56.wav,1,you know what would you do if you had a millionn
```

Note that the last letter in the transcript is repeated to bring it up to the same length as the audio features. Something about these files plus a transcript that is big enough makes the model choke on it.

Some possible fixes for this:

1. Filter the data on `len(features) > len(transcript)`. (Note >, not >=). This is unfortunate because it's not a real requirement, it's just these particular samples that cause trouble, and it could remove good data.
2. Ignore `inf` loss when reducing the mean loss over samples and just pretend the sample doesn't exist. This will fix this problem now but can hide bigger issues in the future.
3. Remove those two files from the dataset.

I'm tempted towards 3 because it's the easiest without collateral damage. @kdavis-mozilla thoughts?

**Comment by kdavis-mozilla** [comment]
I'd say 3 too.

**Comment by reuben** [comment]
@lissyx could you take a look at the test changes I had to make to handle the versioning code?

@kdavis-mozilla things are working and tests are green, so this PR should not change going forward, and I will not force push anymore, so feel free to review.

File: DeepSpeech.py [hunk 16]
```diff
@@ -481,51 +445,51 @@ def train():
                 sys.exit(1)
 
         # Retrieving global_step from restored model and setting training parameters accordingly
-        model_feeder.set_data_set(no_dropout_feed_dict, train_set)
-        step = session.run(global_step, feed_dict=no_dropout_feed_dict)
+        step = session.run(global_step)
         num_gpus = len(Config.available_devices)
-        steps_per_epoch = max(1, train_set.total_batches // num_gpus)
-        steps_trained = step % steps_per_epoch
```
Comment (tilmankamp) [inline DeepSpeech.py]: How do you offset steps of a half-done epoch now? Same question on the feeding side.

File: DeepSpeech.py [hunk 17]
```diff
@@ -481,51 +445,51 @@ def train():
...
         log_debug('target epoch: %d' % target_epoch)
         log_debug('steps per epoch: %d' % steps_per_epoch)
         log_debug('batches per step (GPUs): %d' % num_gpus)
-        log_debug('number of batches in train set: %d' % train_set.total_batches)
-        log_debug('number of batches already trained in epoch: %d' % train_index.index)
+        log_debug('number of batches in train set: %d' % train_batches)
 
-        def run_set(set_name):
-            data_set = getattr(model_feeder, set_name)
+        def run_set(set_name, init_op, num_batches):
```
Comment (tilmankamp) [inline DeepSpeech.py]: nit: Regarding `init_op`: Why not going the dictionary approach - like for the summary writers?

File: util/flags.py [hunk 18]
```diff
@@ -23,6 +23,7 @@ def create_flags():
     # ================
 
     tf.app.flags.DEFINE_boolean ('train',            True,        'whether to train the network')
+    tf.app.flags.DEFINE_boolean ('dev',              True,        'whether to run validation epochs')
```
Comment (tilmankamp) [inline util/flags.py]: I already thought a couple of times about dropping these flags in favour of checking, if the according data set has been provided.

File: DeepSpeech.py [hunk 19]
```diff
@@ -534,68 +498,66 @@ def run_set(set_name):
...
+                        dev_loss = run_set('dev', dev_init_op, dev_batches)
+                        dev_losses.append(dev_loss)
+                        log_info('Finished validating epoch %d - loss: %f' % (current_epoch, dev_loss))
+
+                        if dev_loss < best_dev_loss:
+                            best_dev_loss = dev_loss
+                            save_path = best_dev_saver.save(session, best_dev_path, latest_filename=best_dev_filename)
+                            log_info("Saved new best validating model with loss %f to: %s" % (best_dev_loss, save_path))
+
+                        # Early stopping
```
Comment (tilmankamp) [inline DeepSpeech.py]: Adding it here, as there is no better place: You should add some wording to the early stopping flags about the requirement of having validation activated (as you made it optional).

File: DeepSpeech.py (hunk 17 above)
Comment (reuben) [reply DeepSpeech.py]: To avoid more coupling of different parts of the code via strings.

File: util/flags.py (hunk 18 above)
Comment (reuben) [reply util/flags.py]: Ah, that's a better idea, I'll change it to check for the files.

File: DeepSpeech.py (hunk 19 above)
Comment (reuben) [reply DeepSpeech.py]: The documentation for the early_stop flag already mentions it's tied to validation, but I'll try to clarify it.

File: DeepSpeech.py (hunk 17 above)
Comment (tilmankamp) [reply DeepSpeech.py]: I stated this from a symmetry perspective - either doing it the one or the other way.

File: DeepSpeech.py (hunk 16 above)
Comment (reuben) [reply DeepSpeech.py]: You don't. For a lot of the experiments we've been doing it doesn't make any sense to try and start from the middle of an epoch, since there's no guarantee that the previous training run used the same datasets as the current one. My plan is to follow up on this PR with another one that makes all training runs start from epoch 0, and make the global step be the "version of a checkpoint", which is both more precise and matches the language used by other researchers. Then we can say "release 0.5.0 was trained on this data for 75000 steps" in releases, and there's no confusion of epoch counts varying wildly when the dataset changes.

Comment (reuben) [reply DeepSpeech.py]: See for example:

https://discourse.mozilla.org/t/increasing-train-batch-size-2-to-train-batch-size-3-causes-deepspeech-not-to-train-anymore-why/29790
https://discourse.mozilla.org/t/what-does-the-test-of-epoch-number-mean/29770
https://discourse.mozilla.org/t/problem-in-using-checkpoints-different-datasets-start-with-different-epoch-value/37407
https://discourse.mozilla.org/t/checkpoint-resuming-no-of-epochs-do-not-match-on-google-colab/37358

Comment (tilmankamp) [reply DeepSpeech.py]: A long time ago we already discussed about it, I remember now. Bold move. I also like "release 0.5.0 was trained on this data for 75000 steps".
